pkg/controller/v1alpha2/core/scopes/healthscope/healthscope.go

```diff
@@ -19,133 +19,244 @@ package healthscope
+	}
+	r.IsHealthy = true
+	return r
 }
 
-func resourceHealthStatus(ctx context.Context, client client.Client, namespace string, ref runtimev1alpha1.TypedReference) error {
-	if ref.GroupVersionKind() == apps.SchemeGroupVersion.WithKind("Deployment") {
-		return deploymentHealthStatus(ctx, client, namespace, ref)
+// CheckDaemonsetHealth checks health status of DaemonSet
+func CheckDaemonsetHealth(ctx context.Context, client client.Client, ref runtimev1alpha1.TypedReference, namespace string) *HealthCondition {
+	if ref.GroupVersionKind() != apps.SchemeGroupVersion.WithKind(kindDaemonSet) {
+		return nil
+	}
+	r := &HealthCondition{
+		IsHealthy: false,
+		Target:    ref,
+	}
+	daemonset := apps.DaemonSet{}
+	daemonset.APIVersion = ref.APIVersion
+	daemonset.Kind = ref.Kind
+	nk := types.NamespacedName{Namespace: namespace, Name: ref.Name}
+	if err := client.Get(ctx, nk, &daemonset); err != nil {
+		r.Diagnosis = errors.Wrap(err, errHealthCheck).Error()
+		return r
 	}
+	r.Target.UID = daemonset.GetUID()
 
-	// TODO(artursouza): add other health checks.
-	// Generic health check by validating if the resource exists.
-	object := unstructured.Unstructured{}
-	object.SetAPIVersion(ref.APIVersion)
-	object.SetKind(ref.Kind)
-	objectRef := types.NamespacedName{Namespace: namespace, Name: ref.Name}
-	err := client.Get(ctx, objectRef, &object)
-	return err
+	if daemonset.Status.NumberUnavailable != 0 {
+		r.Diagnosis = fmt.Sprintf(errFmtResourceNotReady, daemonset.Status)
+		return r
+	}
+	r.IsHealthy = true
+	return r
 }
 
-func deploymentHealthStatus(ctx context.Context, client client.Client, namespace string, ref runtimev1alpha1.TypedReference) error {
-	deployment := apps.Deployment{}
-	deployment.APIVersion = ref.APIVersion
-	deployment.Kind = ref.Kind
-	deploymentRef := types.NamespacedName{Namespace: namespace, Name: ref.Name}
-	if err := client.Get(ctx, deploymentRef, &deployment); err != nil {
-		return errors.Wrapf(err, errResourceNotFound, ref.APIVersion, ref.Kind, ref.Name)
+// GeneralHealthChecker checks a list of workload fields
+// if any field check passes, it's a healthy workload
+func GeneralHealthChecker(ctx context.Context, client client.Client, ref runtimev1alpha1.TypedReference, namespace string) *HealthCondition {
+	o := unstructured.Unstructured{}
+	o.SetAPIVersion(ref.APIVersion)
+	o.SetKind(ref.Kind)
+
+	nk := types.NamespacedName{Namespace: namespace, Name: ref.Name}
+	if err := client.Get(ctx, nk, &o); err != nil {
+		return &HealthCondition{
+			IsHealthy: false,
+			Target:    ref,
+			Diagnosis: errors.Wrap(err, errHealthCheck).Error(),
+		}
 	}
+	ref.UID = o.GetUID()
+	pavedV := fieldpath.Pave(o.UnstructuredContent())
 
-	if deployment.Status.ReadyReplicas == 0 {
-		return fmt.Errorf(errDeploymentUnavailable, ref.APIVersion, ref.Kind, ref.Name)
+	for _, fp := range generalReplicaCheckFiled {
+		if value, err := pavedV.GetNumber(fp); err == nil {
+			// just check ready/available replica exists
+			if value != 0 {
```

Comment:
just one replica exists doesn't mean it's healthy...

Reply:
Yes, it seems such kind of general checking rule doesn't make sense... Maybe I should remove `generalReplicaCheck` and leave it to `HealthCheckTrait` to satisfy diverse health checking needs.